src/julia_internal.h

```diff
@@ -590,7 +590,89 @@ jl_tupletype_t *arg_type_tuple(jl_value_t *arg1, jl_value_t **args, size_t nargs
 
 int jl_has_meta(jl_array_t *body, jl_sym_t *sym);
 
-// backtraces
+//--------------------------------------------------
+// Backtraces
+
+// Backtrace buffers:
+//
+// A backtrace buffer conceptually contains a stack of instruction pointers
+// ordered from the inner-most frame to the outermost. We store them in a
+// special raw format for two reasons:
+//
+//   * Efficiency: Every `throw()` must populate the trace so it must be as
+//     efficient as possible.
+//   * Signal safety: For signal-based exceptions such as StackOverflowError
+//     the trace buffer needs to be filled from a signal handler where most
+//     operations are not allowed (including malloc) so we choose a flat
+//     preallocated buffer.
+//
+// The raw buffer layout contains "frame entries" composed of one or several
+// `jl_bt_element_t` values. From the point of view of the GC, an entry is
+// either:
+//
+// 1. A single instruction pointer to native code, not GC-managed.
+// 2. An "extended entry": a mixture of raw data and pointers to julia objects
+//    which must be treated as GC roots.
+//
+// An extended entry `e` is made up of several jl_bt_element_t values:
+//
+//   e[0]  JL_BT_NON_PTR_ENTRY  - Special marker to distinguish extended entries
+//   e[1]  tags                 - A bit packed uintptr_t containing a tag and
+//                                the number of GC- managed and non-managed values
+//   e[2+j]                     - GC managed data
+//   e[2+ngc+i]                 - Non-GC-managed data
+//
+// The format of `tags` is, from LSB to MSB:
+//   0:2     ngc     Number of GC-managed pointers for this frame entry
+//   3:5     nptr    Number of non-GC-managed buffer elements
+//   6:9     tag     Entry type
+//   10:...  header  Entry-specific header data
+typedef struct _jl_bt_element_t {
+    union {
+        uintptr_t   uintptr; // Metadata or native instruction ptr
+        jl_value_t* jlvalue; // Pointer to GC-managed value
+    };
+} jl_bt_element_t;
+
+#define JL_BT_NON_PTR_ENTRY (((uintptr_t)0)-1)
+#define JL_BT_MAX_ENTRY_SIZE 16
+
+STATIC_INLINE int jl_bt_is_native(jl_bt_element_t *bt_entry) JL_NOTSAFEPOINT
+{
+    return bt_entry[0].uintptr != JL_BT_NON_PTR_ENTRY;
+}
+
+// Extended backtrace entry header packing; the bit packing is done manually
+// for precise layout control for interop with julia side.
+STATIC_INLINE uintptr_t jl_bt_pack_extended_entry(int ngc, int nptr,
+                                                  int tag, uintptr_t header) JL_NOTSAFEPOINT
+{
+    assert(((ngc & 0x7) == ngc) && ((nptr & 0x7) == nptr) && ((tag & 0xf) == tag));
+    return (ngc & 0x7) | (nptr & 0x7) << 3 | (tag & 0xf) << 6 | header << 10;
+}
+
+// Unpacking of extended backtrace entry data
+STATIC_INLINE size_t jl_bt_num_jlvals(jl_bt_element_t *bt_entry)      { assert(!jl_bt_is_native(bt_entry)); return  bt_entry[1].uintptr       & 0x7; } JL_NOTSAFEPOINT
```

Comment:
Thanks! I was just now puzzling over what stupid thing I'd done which the analyzer was unhappy with. I'll reformat those and fix the annotations.

Regarding layout, the current `_reformat_bt` unpacks these on the julia side — how do we make sure we're getting the same bits out on the julia side? It seems platform ABIs are free to pack the bitfield fields LSB to MSB or vice versa.

Reply:
They could, but I think we already assume it is little-endian elsewhere. If we encountered a big-endian system, we’d modify the bitfield definitions to still generate the layout we want.

Reply:
I'd love to do this with a bit field, it would remove a lot of boilerplate and be more self-documenting. I went hunting for places where we assume the layout - it looks like we do so in `emit_arrayndims` at least.

Reply:
I tried doing this, but surprisingly it didn't allow any significant boilerplate to be removed so I think I'll leave it as-is. The reason it didn't improve things is that the `bt_entry` pointer is treated conceptually as a pointer to the backtrace entry *as a whole*, which allows most of the code to be unaware that it should access `bt_entry[1]` to get at the sizes and tag (rather than `bt_entry[0]`). This layer of abstraction really helped me get the various code which uses `bt_entry` correct, so I'd like to leave the inline accessor functions as they are.

Reply:
I've done the reformatting and fixed the static analyzer.